d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -534,34 +581,67 @@ private void handleResourceRemoval(Collection<String> removedResources, Resource
     {
       return;
     }
+
+    Map<String, ResourceSubscriber> subscribers = getResourceSubscriberMap(type);
+    WildcardResourceSubscriber wildcardSubscriber = getWildcardResourceSubscriber(type);
     for (String resourceName : removedResources)
     {
       _xdsClientJmx.incrementResourceNotFoundCount();
       _log.warn("Received response that {} {} was removed", type, resourceName);
-      ResourceSubscriber subscriber = getResourceSubscriberMap(type).get(resourceName);
+
+      ResourceSubscriber subscriber = subscribers.get(resourceName);
       if (subscriber != null)
       {
         subscriber.onRemoval();
       }
+
+      if (wildcardSubscriber != null)
+      {
+        wildcardSubscriber.onRemoval(resourceName);
+      }
     }
-    WildcardResourceSubscriber wildcardSubscriber = getWildcardResourceSubscriber(type);
-    if (wildcardSubscriber != null)
+  }
+
+  // If the nonce indicates that this is the last chunk of the response, notify the wildcard subscriber.
+  private void notifyOnLastChunk(DiscoveryResponseData response)
+  {
+    ResourceType type = response.getResourceType();
+    WildcardResourceSubscriber wildcardResourceSubscriber = getWildcardResourceSubscriber(type);
+    if (wildcardResourceSubscriber == null)
     {
-      removedResources.forEach(wildcardSubscriber::onRemoval);
+      return;
     }
-  }
 
+    int remainingChunks;
+    try
+    {
+      byte[] bytes = Hex.decodeHex(response.getNonce().toCharArray());
+      ByteBuffer bb = ByteBuffer.wrap(Arrays.copyOfRange(bytes, 8, 12));
```

Comment:
You don't need to copy this actually, you can use the direct reference here, since you aren't modifying this. Just do:
```suggestion
      byte[] bytes = Hex.decodeHex(response.getNonce().toCharArray());
      ByteBuffer bb = ByteBuffer.wrap(bytes, 8, 4);
```